From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SupeChicken666 <me@supechicken666.dev>
Date: Fri, 25 Sep 2026 03:42:07 +0800
Subject: [PATCH 09/10] audio: Allow customizing nominal latency via property

Set it with ro.boot.audio.nominal_latency_ms (default: 21ms)

Signed-off-by: SupeChicken666 <me@supechicken666.dev>
---
 audio/aidl/default/ModulePrimary.cpp     | 7 +++----
 audio/aidl/default/pulse/StreamPulse.cpp | 9 ++++++---
 2 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/audio/aidl/default/ModulePrimary.cpp b/audio/aidl/default/ModulePrimary.cpp
index ed56cdfbb..91e664e24 100644
--- a/audio/aidl/default/ModulePrimary.cpp
+++ b/audio/aidl/default/ModulePrimary.cpp
@@ -42,6 +42,7 @@ using aidl::android::media::audio::common::AudioPortConfig;
 using aidl::android::media::audio::common::AudioPortExt;
 using aidl::android::media::audio::common::MicrophoneInfo;
 using android::base::GetBoolProperty;
+using android::base::GetIntProperty;
 
 namespace aidl::android::hardware::audio::core {
 
@@ -129,10 +130,8 @@ ndk::ScopedAStatus ModulePrimary::createMmapBuffer(const AudioPortConfig& portCo
     return ndk::ScopedAStatus::ok();
 }
 
-int32_t ModulePrimary::getNominalLatencyMs(const AudioPortConfig& portConfig) {
-    static constexpr int32_t kLowLatencyMs = 10;
-    static constexpr int32_t kStandardLatencyMs = 21;
-    return hasMmapFlag(portConfig.flags.value()) ? kLowLatencyMs : kStandardLatencyMs;
+int32_t ModulePrimary::getNominalLatencyMs(const AudioPortConfig& /* portConfig */) {
+    return GetIntProperty("ro.boot.audio.nominal_latency_ms", 21);
 }
 
 }  // namespace aidl::android::hardware::audio::core
diff --git a/audio/aidl/default/pulse/StreamPulse.cpp b/audio/aidl/default/pulse/StreamPulse.cpp
index 71501c60d..974b7fe36 100644
--- a/audio/aidl/default/pulse/StreamPulse.cpp
+++ b/audio/aidl/default/pulse/StreamPulse.cpp
@@ -84,12 +84,15 @@ StreamPulse::~StreamPulse() {
     }
 
     mPACtx->withLock([&]() {
+        // Calculate buffer size with target latency
+        const uint32_t bufferSize = pa_usec_to_bytes(mContext.getNominalLatencyMs() * PA_USEC_PER_MSEC, &mPASampleSpec);
+
         const pa_buffer_attr bufferAttr = {
-            .maxlength = static_cast<uint32_t>(mBufferSizeFrames * mFrameSizeBytes * 2),
-            .tlength = static_cast<uint32_t>(pa_usec_to_bytes(21000, &mPASampleSpec)),
+            .maxlength = mBufferSizeFrames * mFrameSizeBytes * 2,
+            .tlength = bufferSize,
             .prebuf = static_cast<uint32_t>(-1),
             .minreq = static_cast<uint32_t>(-1),
-            .fragsize = static_cast<uint32_t>(mBufferSizeFrames * mFrameSizeBytes),
+            .fragsize = bufferSize,
         };
 
         constexpr pa_stream_flags_t streamFlags = static_cast<pa_stream_flags_t>(
